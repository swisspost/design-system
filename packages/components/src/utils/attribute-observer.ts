export const getAttributeObserver = (
  attribute: string,
  handler: (element: HTMLElement, mutation?: MutationRecord) => void,
) => {
  /**
   * Handle attribute changes and childList changes from the observer
   * @param {MutationRecord[]} mutationList
   */
  const observerHandler: MutationCallback = (mutationList: MutationRecord[]) => {
    mutationList.forEach(mutation => {
      if (mutation.type === 'attributes' && mutation.attributeName === attribute) {
        handler(mutation.target as HTMLElement);
      }

      if (mutation.type === 'childList') {
        mutation.addedNodes.forEach(node => {
          if (
            node.nodeType === Node.ELEMENT_NODE &&
            (node as HTMLElement).hasAttribute(attribute)
          ) {
            handler(node as HTMLElement);
          }
        });
      }
    });
  };

  // Initialize a mutation observer for patching accessibility features
  return new MutationObserver(observerHandler);
};
