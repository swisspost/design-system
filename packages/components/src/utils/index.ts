export * from './get-element-height';
export * from './on-transition-end';
export * from './property-checkers';
export * from './should-reduce-motion';
