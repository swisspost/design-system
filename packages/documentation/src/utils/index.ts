export * from './component-properties';
export * from './get-attributes';
export * from './map-classes';
export * from './spread-args';
