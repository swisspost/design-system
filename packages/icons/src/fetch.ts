import type { IconSet, SourceReport } from './models/icon.model';
import iconSets from './iconsets.config';
import { checkEnvVarsExist } from './utilities/environment';
import { setup } from './utilities/download/setup';
import { getBaseSourceReport, coloredLogMessage } from './utilities/shared';
import { fetchPage } from './utilities/download/fetchPage';
import { fetchFile } from './utilities/download/fetchFile';
import { format } from './utilities/download/format';
import { updateReport, writeReport } from './utilities/download/report';
import buildSVGs from './utilities/build';

async function fetchSVGs() {
  setup();

  for (const iconSet of iconSets) {
    console.log(coloredLogMessage(`<blue>Start downloading "${iconSet.name}" icons...</blue>`));
    const report = await downloadIconSet(iconSet, getBaseSourceReport());
    console.log(
      coloredLogMessage(
        `<blue>Downloading "${iconSet.name}" finished.</blue>\nDownloaded <green>${report.stats.success}</green> icons, <red>${report.stats.errors}</red> errored, <red>${report.stats.noSVG}</red> with wrong format (no SVG).\n`,
      ),
    );
  }

  buildSVGs();
}

async function downloadIconSet(
  iconSet: IconSet,
  report: SourceReport,
  nextPage?: string,
): Promise<SourceReport> {
  const body = await fetchPage(nextPage ?? iconSet.apiUrl);

  if (body === undefined) {
    throw new Error(`Fetch icons failed, response was ${body}`);
  }

  if ('error' in body) {
    throw new Error(`Fetch icons failed: ${body.error}`);
  }

  console.log(
    coloredLogMessage(
      `<yellow>Fetching icons</yellow> ${body.offset} - ${body.offset + body.count} of ${
        body['total-count']
      }`,
    ),
  );

  await Promise.all(
    format(body).map(async icon => {
      let svg: string | false = false;

      try {
        svg = await fetchFile(icon, iconSet.downloadDirectory);
      } catch (err) {
        console.log(err);
        icon.errored = true;

        if (err instanceof Error) {
          icon.errorMessage = err.message;
        }
      }

      report = updateReport(iconSet, svg, icon, report);
    }),
  );

  if (body.page.next) {
    // Recursively fetch more pages
    return downloadIconSet(iconSet, report, body.page.next);
  }

  return writeReport(iconSet, report);
}

if (checkEnvVarsExist()) {
  fetchSVGs();
}
