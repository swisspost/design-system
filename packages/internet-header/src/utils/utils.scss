@use '@swisspost/design-system-styles/components/fonts';
@use '@swisspost/design-system-styles/variables/color';
@use '@swisspost/design-system-styles/variables/breakpoints';
@use '@swisspost/design-system-styles/variables/type';
@use './mixins.scss';

*,
:host,
*::before,
*::after {
  box-sizing: border-box;
}

:host {
  font-family: type.$font-family-sans-serif;
}

button {
  font: inherit;
  padding: 0;
}

img,
svg {
  max-width: 100%;
  max-height: 100%;
}

svg {
  @media (forced-colors: active) {
    color: white;
  }
}

%no-list {
  list-style: none;
  padding-left: 0;
  margin-top: 0;
  margin-bottom: 0;
}

%blank-button {
  background-color: transparent;
  border: none;
  border-radius: 0;
  padding: 0;
}

%nav-link {
  text-decoration: none;
  color: rgba(color.$black, 0.8);
  transition: color 200ms;
  background-color: transparent;
  border-radius: 0;
  box-shadow: none;
  border: 0;
  margin: 0; // Buttons in safari have a default margin of 2px

  display: flex;
  align-items: center;
  gap: 0.5rem;

  &:hover,
  &:focus {
    color: black;
  }

  > svg {
    width: 1.4em;
    height: 1.4em;
    flex-shrink: 0;
  }

  > span {
    flex-shrink: 1;
  }
}

.box {
  > *:first-child {
    margin-top: 0;
  }

  > *:last-child {
    margin-bottom: 0;
  }
}

.mirrored {
  transform: scaleX(-1);
}

.rotate-270 {
  transform: rotate(270deg);
}

.bold {
  font-weight: 700;
}

.light {
  font-weight: 300;
}

.d-flex {
  display: flex;
}

.d-inline-flex {
  display: inline-flex;
}

.align-items-center {
  align-items: center;
}

.nav-link {
  @extend %nav-link;
}

.no-list {
  @extend %no-list;
}

.btn-blank {
  @extend %blank-button;
}

.wide-container {
  @media (min-width: 1441px) {
    margin: 0 auto;
    max-width: 1440px;
  }
}

.container {
  @include mixins.container-padding();
}

%visually-hidden {
  position: absolute;
  width: 1px;
  height: 1px;
  border: 0;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0 0 0 0);
}

.visually-hidden {
  @extend %visually-hidden;
}

@each $breakpoint, $value in breakpoints.$grid-breakpoints {
  .hidden-#{$breakpoint} {
    @include mixins.max($breakpoint) {
      display: none;
    }
  }
}
