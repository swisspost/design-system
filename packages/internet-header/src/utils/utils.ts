/**
 * Generates a (sequential) unique ID. If prefix is given, the ID is appended to it.
 *
 * @param prefix The value to prefix the ID with
 * @returns Unique ID
 */
export const uniqueId = (() => {
  let num = 0;
  return function (prefix?: string) {
    prefix = prefix || '';
    num += 1;
    return prefix + num;
  };
})();

export const prefersReducedMotion = window.matchMedia('(prefers-reduced-motion: reduce)').matches;
