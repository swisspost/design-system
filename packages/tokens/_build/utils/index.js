export { default as objectDeepmerge } from './object-deepmerge.js';
export { default as objectTextoutput } from './object-textoutput.js';
