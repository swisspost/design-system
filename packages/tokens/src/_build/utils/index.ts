export { default as objectDeepmerge } from './object-deepmerge.ts';
export { default as objectTextoutput } from './object-textoutput.ts';
