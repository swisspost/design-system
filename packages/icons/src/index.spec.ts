import { mocked } from 'ts-jest/utils';
import fetch, { Response } from 'node-fetch';
import { fetchPage, mapResponse } from './index';
import testResult from '../tests/fixtures/test-result.json';
import { CenshareResultPage } from './censhare-result-page';

jest.mock('node-fetch');

describe('fetchPage', () => {
  it('Should fetch a page of data', async () => {
    mocked(fetch).mockImplementationOnce(() =>
      Promise.resolve({ json: () => Promise.resolve(testResult) } as Response),
    );
    const json = await fetchPage('test');
    expect(json?.count).toBeDefined();
    expect(json?.count).toBe(10);
  });

  it('Should return undefined if the call fails', async () => {
    mocked(fetch).mockImplementationOnce(() => Promise.reject('NOK'));
    const json = await fetchPage('test');
    // TODO: check if error was logged correctly
    expect(json).toBe(undefined);
  });
});

describe('mapResponse', () => {
  it('Should map the result object to a meaningful data structure', () => {
    const mapped = mapResponse(testResult as CenshareResultPage);
    expect(mapped).toMatchSnapshot();
  });
});
