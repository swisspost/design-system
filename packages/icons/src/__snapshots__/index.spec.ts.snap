// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapResponse Should map the result object to a meaningful data structure 1`] = `
Array [
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Briefeinwurf, Brief, Kuvert, Briefumschlag, Briefkasten, Briefaufgabe, Boîte aux lettres, Lettre, enveloppe, boîte aux lettres, dépôt de lettres, Letter box, letter, letter cover, envelope, letter box, posting letters, Briefeinwurf, Brief, Kuvert, Briefumschlag, Briefkasten, Briefaufgabe, Boîte aux lettres, Lettre, enveloppe, boîte aux lettres, dépôt de lettres, Letter box, letter, letter cover, envelope, letter box, posting letters",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352084/storage/MzU0NzA1LzAvbWFzdGVy",
    "id": 352084,
    "modifiedAt": 2021-10-18T11:24:43.000Z,
    "name": "1000.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
      "year": "2017",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Brief, Briefaufgabe, Kuvert, Briefumschlag, Lettre, dépôt de lettres, enveloppe, , Letter, posting letters, envelope, Brief, Briefaufgabe, Kuvert, Briefumschlag, Lettre, dépôt de lettres, enveloppe, , Letter, posting letters, envelope",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352086/storage/MzU0NzA2LzAvbWFzdGVy",
    "id": 352086,
    "modifiedAt": 2021-10-18T11:24:40.000Z,
    "name": "1001.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, PostCard Creator, Postkarte, Karte, PostCard Creator, Carte postale, carte, PostCard Creator, postcard, card, PostCard Creator, Postkarte, Karte, PostCard Creator, Carte postale, carte, PostCard Creator, postcard, card",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352087/storage/MzU0NzA3LzAvbWFzdGVy",
    "id": 352087,
    "modifiedAt": 2021-10-18T11:24:40.000Z,
    "name": "1002.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Gratis Poskarte, Geschenkskarte, Postkarte, Karte, Geschenk, Brief, Carte postale gratuite, carte cadeau, Carte postale, carte, cadeau, lettre, Free postcard, gift card, postcard, card, gift, letter, Gratis Poskarte, Geschenkskarte, Postkarte, Karte, Geschenk, Brief, Carte postale gratuite, carte cadeau, Carte postale, carte, cadeau, lettre, Free postcard, gift card, postcard, card, gift, letter",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352088/storage/MzU0NzA4LzAvbWFzdGVy",
    "id": 352088,
    "modifiedAt": 2021-10-18T11:24:41.000Z,
    "name": "1003.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Brief/Paket, Brief, Paket, Kuvert, Paketaufgabe, Briefaufgabe, Briefumschlag, versenden, senden, empfangen, Lettre/colis, Lettre colis, enveloppe, dépôt de colis, dépôt de lettres, enveloppe, envoyer, , Letter/parcel, letter, parcel, letter cover, posting parcels, posting letters, envelope, dispatch, send, receive, Brief/Paket, Brief, Paket, Kuvert, Paketaufgabe, Briefaufgabe, Briefumschlag, versenden, senden, empfangen, Lettre/colis, Lettre colis, enveloppe, dépôt de colis, dépôt de lettres, enveloppe, envoyer, , Letter/parcel, letter, parcel, letter cover, posting parcels, posting letters, envelope, dispatch, send, receive",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352089/storage/MzU0NzA5LzAvbWFzdGVy",
    "id": 352089,
    "modifiedAt": 2021-10-18T11:24:41.000Z,
    "name": "1004.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Paket, Paketaufgabe, Geschenk, Schachtel, senden, empfangen, Colis, dépôt de colis, Cadeau, boîte, envoyer, recevoir, Parcel, posting parcels, gift, box, send, receive, Paket, Paketaufgabe, Geschenk, Schachtel, senden, empfangen, Colis, dépôt de colis, Cadeau, boîte, envoyer, recevoir, Parcel, posting parcels, gift, box, send, receive",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352090/storage/MzU0NzEwLzAvbWFzdGVy",
    "id": 352090,
    "modifiedAt": 2021-10-18T11:24:42.000Z,
    "name": "1005.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, My Post 24, Paket, 24h, Paketaufgabe, senden, empfangen, My Post 24, Colis, 24h, dépôt de colis, envoyer, recevoir, My Post 24, parcel, 24h, sending parcels, send, receive, My Post 24, Paket, 24h, Paketaufgabe, senden, empfangen, My Post 24, Colis, 24h, dépôt de colis, envoyer, recevoir, My Post 24, parcel, 24h, sending parcels, send, receive",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352091/storage/MzU0NzExLzAvbWFzdGVy",
    "id": 352091,
    "modifiedAt": 2021-10-18T11:16:36.000Z,
    "name": "1006.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
      "year": "2017",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Zustellservice, Zusteller, Paket, Brief, Mensch, Empfänger, Sender, Post empfangen, Paket und Brief, Service de distribution, Personnel de distribution, colis, lettre, homme, destinataire, expéditeur, envoyer du courrier, colis et lettre, Delivery service, delivery staff, parcel, letter, human, recipient, sender, receiving mail, parcel and letter, Zustellservice, Zusteller, Paket, Brief, Mensch, Empfänger, Sender, Post empfangen, Paket und Brief, Service de distribution, Personnel de distribution, colis, lettre, homme, destinataire, expéditeur, envoyer du courrier, colis et lettre, Delivery service, delivery staff, parcel, letter, human, recipient, sender, receiving mail, parcel and letter",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352092/storage/MzU0NzEyLzAvbWFzdGVy",
    "id": 352092,
    "modifiedAt": 2021-10-18T11:24:42.000Z,
    "name": "1007.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Briefzentrum, Zustellung, Zuhause, Brief, Kuvert, Briefe empfangen, Briefumschlag, Centre courrier, Distribution, domicile, lettre, enveloppe, recevoir des lettres, enveloppe, Letter center, delivery, home, letter, envelope, receiving letters, letter cover",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352093/storage/MzU0NzEzLzAvbWFzdGVy",
    "id": 352093,
    "modifiedAt": 2021-10-18T11:32:39.000Z,
    "name": "1008.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
  Object {
    "contentInfo": Object {
      "freeKeywords": "Piktogramme \\"Die Post\\" ab 2017, Paketzentrum, Zustellung, Zuhause, Paket, empfangen, versenden, senden, Centre colis, Distribution, domicile, colis, recevoir, envoyer, expédier, delivery, at home, parcel, receive, dispatch, send",
    },
    "downloadLink": "https://cdn.post.ch/hcms/v2.0/entity/asset/352094/storage/MzU0NzE0LzAvbWFzdGVy",
    "id": 352094,
    "modifiedAt": 2021-10-18T11:32:38.000Z,
    "name": "1009.svg",
    "postInfo": Object {
      "businessfield": "kommunikation",
    },
    "type": "picture.pictogram.",
    "typeFilter": "pictograms",
  },
]
`;
